<a href="https://colab.research.google.com/github/MattWroclaw/data-science-bootcamp/blob/main/Uczenie_Glebokie/02_keras.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Tensorflow
>Strona biblioteki: [https://www.tensorflow.org/](https://www.tensorflow.org/)  
>Dokumentacja: [https://www.tensorflow.org/api_docs](https://www.tensorflow.org/api_docs)  
>
>Szybko rozwijająca się biblioteka do uczenia maszynowego i uczenia głębokiego rozwijana przez Google.
>
>Aby zainstalować bibliotekę tensorflow, użyj polecenia poniżej:
```
pip install tensorflow
```

### Keras
>Strona biblioteki: [https://keras.io/](https://keras.io/)  
>
>Wysokopoziomowy interfejs do budowy sieci neuronowych. Działa zarówno na Tensorflow, Theano oraz CNTK.
>
>Aby zainstalować bibliotekę keras, użyj polecenia poniżej:
```
pip install keras
```

### Model sekwencjny (Sequential Model):
1. [Wprowadzenie - pierwszy model](#a0)
2. [Funkcje aktywacji](#a1)
3. [Kompilacja modelu](#a2)
4. [Trenowanie modelu](#a3)
5. [Przykład - klasyfikacja binarna](#a4)
6. [Przykład - klasyfikacja wieloklasowa](#a5)
7. [Przykład - regresja](#a6)



Załadowanie biblioteki tensorflow


In [1]:
%tensorflow_version 2.x
import tensorflow as tf
import numpy as np
import pandas as pd
import plotly.express as px
tf.__version__

Colab only includes TensorFlow 2.x; %tensorflow_version has no effect.


'2.17.0'

### <a name='a0'></a>Wprowadzenie - pierwszy model
Model sekwencyjny to nic innego jak liniowy stos warstw.


In [2]:
# utworzenie instancji klasy Sequential
from tensorflow.keras.models import Sequential

model = Sequential()
print(model)

<Sequential name=sequential, built=False>


Podstawowym elementem składowym modelu są warstwy. Aby dodać najbardziej standardową warstwę - warstwę gęsto połączoną należy użyć warstwy **Dense**. Aby dodać warstwę do modelu należy użyć metody *.add()*

In [3]:
from tensorflow.keras.layers import Dense

model.add(Dense(units=4, input_shape=(10,)))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Wyświetlenie podsumowania modelu: metoda *.summary()*

Kod ten definiuje warstwę w sieci neuronowej z wykorzystaniem biblioteki Keras w TensorFlow. Oto jego szczegóły:

### Składnia:
```python
from tensorflow.keras.layers import Dense

model.add(Dense(units=4, input_shape=(10,)))
```

### Wyjaśnienie:
1. **`Dense`**:
   - **`Dense`** to rodzaj w pełni połączonej (fully connected) warstwy w sieciach neuronowych. Oznacza to, że każdy neuron w tej warstwie jest połączony z każdym neuronem w poprzedniej warstwie.
   - W pełni połączona warstwa oblicza liniową kombinację wartości wejściowych, a następnie stosuje funkcję aktywacji (domyślnie, jest to `linear`, ale można ustawić inną funkcję aktywacji jak np. `relu`).

2. **Parametry `Dense`**:
   - **`units=4`**:
     - Określa liczbę neuronów w tej warstwie. W tym przypadku warstwa ma 4 neurony.
     - Każdy neuron w tej warstwie otrzyma połączenie z każdym z neuronów z poprzedniej warstwy.
   
   - **`input_shape=(10,)`**:
     - Określa kształt danych wejściowych do sieci. W tym przypadku, każdy przykładowy wektor danych wejściowych ma 10 cech (wartości).
     - `input_shape=(10,)` oznacza, że na wejście tej warstwy trafia 10-wymiarowy wektor (tablica liczb o długości 10).
     - To jest wymagane tylko dla pierwszej warstwy w modelu, ponieważ kolejne warstwy automatycznie określą kształt danych wejściowych na podstawie poprzednich warstw.

### Przykład działania:
Załóżmy, że dane wejściowe do sieci to wektor o 10 elementach, np. `[x1, x2, ..., x10]`. Każdy z tych elementów jest połączony z każdym z 4 neuronów w warstwie. Każdy neuron oblicza liniową kombinację tych 10 wejść i na tej podstawie generuje wynik, który może zostać poddany funkcji aktywacji.

### Działanie krok po kroku:
1. **Dane wejściowe** mają kształt `(10,)`, co oznacza, że każdy przykład składa się z wektora o 10 liczbach.
2. **Warstwa `Dense`** ma 4 neurony. Każdy z tych neuronów połączy się z każdym z 10 wejściowych neuronów, co oznacza, że będzie 40 wag (10 wejść × 4 neurony).
3. Wynikiem działania tej warstwy będzie wektor wyjściowy o rozmiarze 4, który zostanie przekazany do kolejnej warstwy sieci lub jako wyjście modelu (jeśli to ostatnia warstwa).

W ten sposób, ten fragment kodu definiuje prostą w pełni połączoną warstwę z 4 neuronami, która przyjmuje dane wejściowe w postaci wektora o 10 elementach.

In [4]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 4)                   │              44 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 44 (176.00 B)

 Trainable params: 44 (176.00 B)

 Non-trainable params: 0 (0.00 B)

### <a name='a1'></a>Funkcje aktywacji
Istotnym elemenem sieci neuronowych jest dobór odpowiednich funkcji aktywacji. Funkcje aktywacji jak sama nazwa wskazuje są odpowiedzialne za aktywowanie odpowiednich neuronów podczas procesu uczenia.

Jeżeli nie określimy podczas dodawania warstwy funkcji aktywacji, domyślnie stosowana jest liniowa funkcja aktywacji, tzn. $a(x)=x$

Warstwa z liniową funkcją aktywacji może uczyć się tylko liniowych przekształceń danych wejściowych. Dlatego stosuje się różne funkcje aktywacji aby rozwiazywać problemy nieliniowe.

[Keras: Funkcje aktywacji](https://keras.io/activations/)

In [6]:
from tensorflow.keras.activations import linear

random_data = np.linspace(start=-3, stop=3, num=300)
data = pd.DataFrame({'data': random_data, 'linear': linear(random_data)})
data.head()

data    linear
0 -3.000000 -3.000000
1 -2.979933 -2.979933
2 -2.959866 -2.959866
3 -2.939799 -2.939799
4 -2.919732 -2.919732

Ten kod ilustruje, jak działa funkcja aktywacji **`linear`** w Keras i jak dane są przetwarzane i przechowywane w ramce danych (`DataFrame`) przy użyciu Pandas. Oto dokładne wyjaśnienie każdego elementu:

### Składnia:

```python
from tensorflow.keras.activations import linear
random_data = np.linspace(start=-3, stop=3, num=300)
data = pd.DataFrame({'data': random_data, 'linear': linear(random_data)})
data.head()
```

### Wyjaśnienie:

1. **`from tensorflow.keras.activations import linear`**:
   - Importuje funkcję aktywacji **`linear`** z Keras. Funkcja `linear` jest bardzo prosta – zwraca wartość wejściową bez żadnych zmian (jest to po prostu funkcja tożsamościowa: \( f(x) = x \)).
   - W sieciach neuronowych taka funkcja aktywacji nie wprowadza żadnej nieliniowości, więc jest rzadko używana w warstwach ukrytych, ale może być stosowana w warstwach wyjściowych.

2. **`random_data = np.linspace(start=-3, stop=3, num=300)`**:
   - Używa funkcji **`linspace`** z biblioteki **NumPy**, aby wygenerować 300 równomiernie rozmieszczonych wartości pomiędzy -3 a 3.
   - `start=-3` oznacza początkową wartość przedziału, `stop=3` to końcowa wartość, a `num=300` wskazuje, że zostanie wygenerowanych dokładnie 300 liczb.
   - Wartości te zostają zapisane w zmiennej **`random_data`**.

3. **`data = pd.DataFrame({'data': random_data, 'linear': linear(random_data)})`**:
   - Tworzy ramkę danych (**DataFrame**) z biblioteki **Pandas**.
   - Kolumna o nazwie **`data`** zawiera wartości z wektora **`random_data`** (czyli 300 wartości z przedziału od -3 do 3).
   - Kolumna o nazwie **`linear`** zawiera wyniki zastosowania funkcji aktywacji **`linear`** na tych danych wejściowych. W tym przypadku funkcja `linear` nie zmienia wartości, więc kolumna **`linear`** będzie identyczna jak kolumna **`data`**.

4. **`data.head()`**:
   - Wyświetla pierwsze 5 wierszy ramki danych (**DataFrame**) przy pomocy funkcji **`head()`**. Pozwala to szybko zobaczyć przykładowe wartości w ramce danych.

### Działanie krok po kroku:
1. Generujesz 300 liczb z przedziału od -3 do 3 (równo rozłożonych).
2. Tworzysz ramkę danych, która zawiera dwie kolumny:
   - **`data`** – oryginalne wygenerowane liczby.
   - **`linear`** – wynik działania funkcji aktywacji `linear` na tych danych (w tym przypadku będą to te same liczby, ponieważ funkcja `linear` zwraca wartość bez zmian).
3. Wyświetlasz pierwsze 5 wierszy, aby zobaczyć, jak wyglądają dane.

### Przykładowy wynik:
Dla danych z przedziału -3 do 3, przykładowa tabela danych mogłaby wyglądać tak:

| data | linear |
|------|--------|
| -3.0 | -3.0   |
| -2.98| -2.98  |
| -2.96| -2.96  |
| -2.94| -2.94  |
| -2.92| -2.92  |

Obie kolumny są identyczne, ponieważ funkcja aktywacji **`linear`** nie zmienia wartości wejściowej.

In [7]:
px.line(data, x='data', y='linear', width=500, height=400, range_y=[-3, 3])

In [8]:
from tensorflow.keras.activations import sigmoid

data = pd.DataFrame({'data': random_data, 'sigmoid': sigmoid(random_data)})
data.head()

data   sigmoid
0 -3.000000  0.047426
1 -2.979933  0.048341
2 -2.959866  0.049272
3 -2.939799  0.050221
4 -2.919732  0.051187

In [9]:
px.line(data, x='data', y='sigmoid', width=500, height=400, range_y=[-0.5, 1.5])

In [10]:
from tensorflow.keras.activations import relu

data = pd.DataFrame({'data': random_data, 'relu': relu(random_data)})
data.head()

data  relu
0 -3.000000   0.0
1 -2.979933   0.0
2 -2.959866   0.0
3 -2.939799   0.0
4 -2.919732   0.0

In [11]:
px.line(data, x='data', y='relu', width=500, height=400, range_y=[-0.5, 1.5])

In [12]:
from tensorflow.keras.activations import tanh

data = pd.DataFrame({'data': random_data, 'tanh': tanh(random_data)})
data.head()

data      tanh
0 -3.000000 -0.995055
1 -2.979933 -0.994853
2 -2.959866 -0.994643
3 -2.939799 -0.994424
4 -2.919732 -0.994196

In [13]:
px.line(data, x='data', y='tanh', width=500, height=400, range_y=[-1.5, 1.5])

In [14]:
model = Sequential()
model.add(Dense(units=8, activation='relu', input_shape=(10,)))
model.add(Dense(units=1, activation='sigmoid'))
model.summary()

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning:

Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.



Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_1 (Dense)                      │ (None, 8)                   │              88 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 1)                   │               9 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 97 (388.00 B)

 Trainable params: 97 (388.00 B)

 Non-trainable params: 0 (0.00 B)

### <a name='a2'></a>Kompilacja modelu
Przed rozpoczęciem trenowania sieci należy odpowiednio skonfigurować proces uczenia. W tym kroku określamy:
* rodzaj optymalizatora ([Keras - Optymalizatory](https://keras.io/optimizers/))
* funkcję straty ([Keras - Funkcje Straty](https://keras.io/losses/))
* metryki, które będziemy obserwować podczas trenowania sieci ([Keras - Metryki](https://keras.io/metrics/))

In [15]:

# klasyfikacja binarna
model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

# klasyfikacja wieloklasowa
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# regresja
model.compile(optimizer='rmsprop',
              loss='mse')

### <a name='a3'></a>Trenowanie modelu
Za dane wejściowe do modelu należy przekazać Numpy arrays:
* **epochs** - krotność przejścia danych przez sieć w procesie uczenia
* **batch_size** - rozmiar wsadu po którym następuje aktualizacja wag
* **validation_split** - część danych treningowych, które zostaną wykorzystane jako zbiór walidacyjny
* **validation_data** - (x_val, y_val) - dane wykorzystane do walidacji modelu

In [ ]:
# model.fit(data, labels, epochs=10, batch_size=32)
# model.fit(data, labels, epochs=10, batch_size=32, validation_split=0.2)
# model.fit(data, labels, epochs=10, batch_size=32, validation_data=(x_val, y_val))

### <a name='a4'></a> Przykład - klasyfikacja binarna

In [16]:
data = np.random.randn(10000, 150)
labels = np.random.randint(2, size=(10000, 1))

print(data.shape)
print(labels.shape)

(10000, 150)
(10000, 1)


Ten kod generuje losowe dane oraz etykiety i drukuje ich rozmiary (wymiary). Oto dokładne wyjaśnienie:

### Składnia:

```python
data = np.random.randn(10000, 150)
labels = np.random.randint(2, size=(10000, 1))

print(data.shape)
print(labels.shape)
```

### Wyjaśnienie:

1. **`data = np.random.randn(10000, 150)`**:
   - **`np.random.randn`** generuje losowe liczby z rozkładu normalnego (ze średnią 0 i odchyleniem standardowym 1).
   - Tworzy macierz o wymiarach **(10000, 150)**, czyli 10 tysięcy próbek, z których każda ma 150 cech (wartości).
     - **10000**: Liczba próbek.
     - **150**: Liczba cech (kolumn) dla każdej próbki.
   - Wynikiem jest macierz **`data`** o wymiarze **(10000, 150)**, w której każda komórka zawiera losową liczbę zgodną z rozkładem normalnym.

2. **`labels = np.random.randint(2, size=(10000, 1))`**:
   - **`np.random.randint(2, size=(10000, 1))`** generuje losowe liczby całkowite z przedziału **[0, 2)** (czyli wartości 0 lub 1).
   - Tworzy macierz etykiet **`labels`** o wymiarze **(10000, 1)**, gdzie każda próbka ma przypisaną jedną etykietę. Etykiety te to liczby 0 lub 1, co sugeruje, że mogą być to etykiety dla problemu binarnej klasyfikacji (np. klasyfikacja na dwie kategorie).
     - **10000**: Liczba etykiet, jedna dla każdej próbki.
     - **1**: Oznacza, że każda próbka ma przypisaną pojedynczą etykietę (dla problemu binarnej klasyfikacji).

3. **`print(data.shape)`**:
   - Funkcja **`data.shape`** zwraca kształt (wymiary) macierzy **`data`**.
   - W tym przypadku wynik to **`(10000, 150)`**, co oznacza, że macierz **`data`** ma 10 tysięcy wierszy (próbek) i 150 kolumn (cech).

4. **`print(labels.shape)`**:
   - Funkcja **`labels.shape`** zwraca kształt macierzy **`labels`**.
   - W tym przypadku wynik to **`(10000, 1)`**, co oznacza, że macierz **`labels`** ma 10 tysięcy wierszy (próbek) i 1 kolumnę (etykietę dla każdej próbki).

### Co ten kod robi:
- Tworzy dane wejściowe **`data`** składające się z 10 tysięcy próbek, gdzie każda próbka ma 150 cech.
- Tworzy macierz etykiet **`labels`**, która zawiera 10 tysięcy etykiet binarnych (0 lub 1) przypisanych do każdej próbki.
- Wydrukuje rozmiary macierzy, co pozwala upewnić się, że dane mają odpowiednie wymiary.

### Przykładowy wynik:

```python
(10000, 150)
(10000, 1)
```

Ten wynik oznacza, że:
- Macierz danych wejściowych **`data`** ma 10000 wierszy (próbek) i 150 kolumn (cech).
- Macierz etykiet **`labels`** ma 10000 wierszy (próbek) i 1 kolumnę (etykieta dla każdej próbki).


In [17]:
data[:3]

array([[ 1.98685439,  0.06183068,  0.41524232,  0.10279082,  1.48984523,
        -0.3518733 , -0.46992647,  0.95910458,  0.59321826,  0.21345723,
        -0.1325323 , -0.09570502,  0.54615412, -0.40148914,  0.98583442,
        -0.17648368, -0.26242034,  1.00594788,  0.58117512,  1.69219103,
        -0.54380344, -0.04875768,  1.04856737, -0.06867317,  0.59989487,
         0.68942122,  0.3327151 , -0.8034428 , -0.02898505,  0.35960612,
         0.09994916,  0.48776806,  0.50011631, -1.71114287, -0.63225123,
        -0.19252555, -0.82082148, -1.07710525,  2.10236367,  0.94876457,
         0.75921207, -0.63001697,  1.45178822, -0.59122484, -0.13301484,
        -0.55574995,  0.18249052,  0.77440486, -1.06291581,  0.67155827,
         0.60960832,  1.49231952,  0.57640813,  0.60196296, -0.80741091,
         0.94131281,  1.16328725,  0.82703347, -2.40142117, -0.21650208,
         0.95286682, -0.8792608 , -0.37362774,  0.1055718 , -0.86450995,
         1.45251974,  1.34363345,  1.33592289,  0.2

In [18]:
labels[:10]

array([[0],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0]])

In [19]:
model = Sequential()
model.add(Dense(units=32, activation='relu', input_shape=(150,)))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(data, labels, epochs=20)

Epoch 1/20
313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.5014 - loss: 0.7666
Epoch 2/20
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.5445 - loss: 0.6948
Epoch 3/20
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.5722 - loss: 0.6764
Epoch 4/20
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.5992 - loss: 0.6578
Epoch 5/20
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.6287 - loss: 0.6450
Epoch 6/20
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.6580 - loss: 0.6279
Epoch 7/20
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.6705 - loss: 0.6169
Epoch 8/20
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.6899 - loss: 0.6029
Epoch 9/20
313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7038 - loss: 0.5862
Epoch 10/20
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7045 - loss: 0.5789
Epoch 11/20
313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7195 - loss: 0.5651
Epoch 12/20
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step

In [21]:
model = Sequential()
model.add(Dense(units=32, activation='relu', input_shape=(150,)))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(data, labels, epochs=20, batch_size=64)
# zmiana model.fit(data, labels, epochs=20, batch_size=64)

Epoch 1/20
157/157 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.4897 - loss: 0.7746
Epoch 2/20
157/157 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5366 - loss: 0.7012
Epoch 3/20
157/157 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5679 - loss: 0.6816
Epoch 4/20
157/157 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5937 - loss: 0.6697
Epoch 5/20
157/157 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6171 - loss: 0.6570
Epoch 6/20
157/157 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6359 - loss: 0.6437
Epoch 7/20
157/157 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6465 - loss: 0.6357
Epoch 8/20
157/157 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6520 - loss: 0.6267
Epoch 9/20
157/157 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6774 - loss: 0.6137
Epoch 10/20
157/157 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6792 - loss: 0.6055
Epoch 11/20
157/157 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.6939 - loss: 0.5956
Epoch 12/20
157/157 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step

In [22]:
model = Sequential()
model.add(Dense(units=32, activation='relu', input_shape=(150,)))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(data, labels, epochs=20, batch_size=32, validation_split=0.2)

Epoch 1/20
250/250 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.4954 - loss: 0.8217 - val_accuracy: 0.4910 - val_loss: 0.7330
Epoch 2/20
250/250 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5539 - loss: 0.6963 - val_accuracy: 0.4905 - val_loss: 0.7294
Epoch 3/20
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.5989 - loss: 0.6663 - val_accuracy: 0.4930 - val_loss: 0.7316
Epoch 4/20
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6115 - loss: 0.6537 - val_accuracy: 0.4840 - val_loss: 0.7385
Epoch 5/20
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6350 - loss: 0.6397 - val_accuracy: 0.4860 - val_loss: 0.7410
Epoch 6/20
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6651 - loss: 0.6236 - val_accuracy: 0.4900 - val_loss: 0.7441
Epoch 7/20
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6852 - loss: 0.6061 - val_accuracy: 0.4830 - val_loss: 0.7507
Epoch 8/20
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6988 - loss: 0.5867 - val_accuracy: 0.

In [23]:
model = Sequential()
model.add(Dense(units=32, activation='relu', input_shape=(150,)))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

history = model.fit(data, labels, epochs=20, batch_size=32, validation_split=0.2, verbose=0)

In [24]:
metrics = history.history
metrics.keys()

dict_keys(['accuracy', 'loss', 'val_accuracy', 'val_loss'])

In [25]:
model = Sequential()
model.add(Dense(units=32, activation='relu', input_shape=(150,)))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

history = model.fit(data, labels, epochs=20, batch_size=32, validation_split=0.2, verbose=1)

Epoch 1/20
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.5072 - loss: 0.7737 - val_accuracy: 0.4950 - val_loss: 0.7243
Epoch 2/20
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.5502 - loss: 0.6900 - val_accuracy: 0.4950 - val_loss: 0.7215
Epoch 3/20
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.5884 - loss: 0.6701 - val_accuracy: 0.5050 - val_loss: 0.7206
Epoch 4/20
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6086 - loss: 0.6582 - val_accuracy: 0.5050 - val_loss: 0.7246
Epoch 5/20
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6417 - loss: 0.6378 - val_accuracy: 0.5005 - val_loss: 0.7305
Epoch 6/20
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6575 - loss: 0.6282 - val_accuracy: 0.5040 - val_loss: 0.7342
Epoch 7/20
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6756 - loss: 0.6118 - val_accuracy: 0.5045 - val_loss: 0.7417
Epoch 8/20
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7004 - loss: 0.5969 - val_accuracy: 0.

In [26]:
test_data = np.random.randn(5, 150)
test_labels = np.random.randint(2, size=(5, 1))

In [27]:
model.predict(test_data)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 77ms/step


array([[0.3847248 ],
       [0.32524294],
       [0.85316056],
       [0.21549998],
       [0.72550935]], dtype=float32)

In [28]:
model.predict_proba(test_data)

AttributeError: 'Sequential' object has no attribute 'predict_proba'

In [29]:
model.predict_classes(test_data)

AttributeError: 'Sequential' object has no attribute 'predict_classes'

### <a name='a5'></a> Przykład - klasyfikacja wieloklasowa

In [30]:
data = np.random.random((10000, 150))
labels = np.random.randint(10, size=(10000, 1))

In [31]:
data = np.random.random((10000, 150))
labels = np.random.randint(10, size=(10000, 1))

In [32]:
labels[:10]


array([[1],
       [8],
       [1],
       [6],
       [9],
       [8],
       [3],
       [0],
       [4],
       [0]])

In [33]:
from tensorflow.keras.utils import to_categorical
labels = to_categorical(labels, num_classes=10)
labels

array([[0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

Ten kod wykorzystuje funkcję **`to_categorical`** z Keras do przekształcenia etykiet w format "one-hot encoding", co jest powszechnie stosowane w zadaniach klasyfikacji wieloklasowej.

### Składnia:

```python
from tensorflow.keras.utils import to_categorical
labels = to_categorical(labels, num_classes=10)
```

### Wyjaśnienie:

1. **`from tensorflow.keras.utils import to_categorical`**:
   - Importuje funkcję **`to_categorical`** z Keras, która służy do przekształcania wektora etykiet całkowitych (np. 0, 1, 2, ...) na macierz z reprezentacją "one-hot encoding".

2. **`to_categorical(labels, num_classes=10)`**:
   - **`labels`** to tablica lub wektor zawierający etykiety, które są liczbami całkowitymi reprezentującymi klasy. Funkcja przekształca te liczby w format "one-hot encoding".
   - **`num_classes=10`** określa, że mamy 10 możliwych klas (0-9). Każda etykieta będzie zamieniona na wektor o długości 10, gdzie odpowiednia pozycja odpowiadająca danej klasie jest ustawiona na 1, a pozostałe pozycje na 0.
   
   Przykładowo, jeśli etykieta to **2**, to po przekształceniu będzie wyglądała tak:
   ```
   [0, 0, 1, 0, 0, 0, 0, 0, 0, 0]
   ```
   A etykieta **5** zostanie zamieniona na:
   ```
   [0, 0, 0, 0, 0, 1, 0, 0, 0, 0]
   ```

3. **Zastosowanie "one-hot encoding"**:
   - **One-hot encoding** jest często używany w sieciach neuronowych do reprezentacji etykiet klasowych, ponieważ model zwykle przewiduje prawdopodobieństwo przynależności do każdej klasy. One-hot encoding zamienia liczby całkowite (które mogą reprezentować klasy) na binarne wektory, co jest bardziej odpowiednie do pracy z algorytmami uczenia maszynowego, szczególnie w warstwach wyjściowych sieci neuronowych.

### Działanie krok po kroku:
1. **Przed przekształceniem**:
   - Załóżmy, że masz etykiety w formie liczb całkowitych, np.: `[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]`.

2. **Po zastosowaniu `to_categorical(labels, num_classes=10)`**:
   - Etykiety zostaną przekształcone w tablicę o rozmiarze **(liczba próbek, 10)**, gdzie każda próbka będzie reprezentowana jako wektor one-hot:
   ```python
   [
     [1, 0, 0, 0, 0, 0, 0, 0, 0, 0],  # Etykieta 0
     [0, 1, 0, 0, 0, 0, 0, 0, 0, 0],  # Etykieta 1
     [0, 0, 1, 0, 0, 0, 0, 0, 0, 0],  # Etykieta 2
     [0, 0, 0, 1, 0, 0, 0, 0, 0, 0],  # Etykieta 3
     ...
     [0, 0, 0, 0, 0, 0, 0, 0, 0, 1]   # Etykieta 9
   ]
   ```

3. **Po przekształceniu**:
   - W wyniku działania funkcji **`to_categorical`**, każda etykieta jest teraz reprezentowana jako wektor o długości 10, co umożliwia sieci neuronowej przewidywanie prawdopodobieństw dla każdej z 10 klas.

### Dlaczego to jest ważne?
W sieciach neuronowych, w szczególności w problemach klasyfikacji wieloklasowej, model często używa funkcji aktywacji **softmax** na warstwie wyjściowej, która zwraca prawdopodobieństwo przynależności do każdej z klas. **One-hot encoding** pomaga w porównywaniu przewidywań modelu z rzeczywistymi etykietami podczas obliczania funkcji kosztu.

In [34]:
labels[1]

array([0., 0., 0., 0., 0., 0., 0., 0., 1., 0.])

In [35]:
model = Sequential()
model.add(Dense(units=32, activation='relu', input_shape=(150,)))
model.add(Dense(units=10, activation='softmax'))

model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

model.fit(data, labels, batch_size=32, epochs=30, validation_split=0.2)

Epoch 1/30


/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning:

Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.



250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.0983 - loss: 2.3410 - val_accuracy: 0.0955 - val_loss: 2.3097
Epoch 2/30
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.0990 - loss: 2.3048 - val_accuracy: 0.0905 - val_loss: 2.3070
Epoch 3/30
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.1121 - loss: 2.3004 - val_accuracy: 0.0955 - val_loss: 2.3063
Epoch 4/30
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.1217 - loss: 2.2961 - val_accuracy: 0.0870 - val_loss: 2.3084
Epoch 5/30
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.1157 - loss: 2.2950 - val_accuracy: 0.0860 - val_loss: 2.3176
Epoch 6/30
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.1189 - loss: 2.2916 - val_accuracy: 0.0870 - val_loss: 2.3224
Epoch 7/30
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.1283 - loss: 2.2866 - val_accuracy: 0.0915 - val_loss: 2.3203
Epoch 8/30
250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.1321 - loss: 2.2792 - val_accuracy: 0.0930 - val_

In [36]:
test_data = np.random.random((10, 150))

model.predict(test_data)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 43ms/step


array([[0.10719407, 0.1035215 , 0.14556146, 0.02840328, 0.23529303,
        0.09801079, 0.11303296, 0.0471069 , 0.03415169, 0.08772424],
       [0.08823883, 0.06079516, 0.13821141, 0.04690433, 0.21948771,
        0.04699683, 0.07606985, 0.10547666, 0.06830394, 0.1495152 ],
       [0.11025117, 0.05138603, 0.14514637, 0.10930527, 0.10172184,
        0.08792102, 0.12745748, 0.084739  , 0.10639901, 0.07567275],
       [0.12600973, 0.02791144, 0.34069645, 0.07058702, 0.06469314,
        0.06865855, 0.10706916, 0.06802886, 0.07188276, 0.05446288],
       [0.13953291, 0.03260254, 0.2079232 , 0.08324022, 0.12952396,
        0.09240351, 0.10984238, 0.07128226, 0.08482215, 0.04882689],
       [0.10791879, 0.05993644, 0.10183918, 0.10043182, 0.08946417,
        0.06675646, 0.07164737, 0.10470901, 0.18751511, 0.10978165],
       [0.23099135, 0.03995658, 0.17065649, 0.03380148, 0.06173046,
        0.04754536, 0.17414835, 0.13093701, 0.06639566, 0.04383725],
       [0.14466253, 0.02581259, 0.1987838

In [37]:
model.predict_classes(test_data)

AttributeError: 'Sequential' object has no attribute 'predict_classes'